### PR TITLE
feat: New Module `avm/res/network/azure-firewall`, migrated from CARML.

### PR DESCRIPTION
**Pull request by rahalan**
## Description

Migration of Azure firewall module from CARML

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.network.azure-firewall](https://github.com/rahalan/bicep-registry-modules/actions/workflows/avm.res.network.azure-firewall.yml)      |

## Type of Change

- [X]  I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

File: avm/res/compute/virtual-machine/README.md [hunk 1]
```diff
@@ -3646,7 +3646,7 @@ extensionDomainJoinConfig: {
     }
 }
 resource kv1 'Microsoft.KeyVault/vaults@2019-09-01' existing = {
-  name: 'adp-[[namePrefix]]-az-kv-x-001'
+  name: 'adp-#_namePrefix_#-az-kv-x-001'
```
Comment (eriqua) [inline avm/res/compute/virtual-machine/README.md:3649]: Please address any change to different modules than azure firewall in a new PR

Comment (rahalan) [reply avm/res/compute/virtual-machine/README.md:3649]: @eriqua new PR is created: https://github.com/Azure/bicep-registry-modules/pull/1096

Comment (rahalan) [reply avm/res/compute/virtual-machine/README.md]: @eriqua Can we resolve the conversation, as the PR #1096 made that changes, so they are already in main and thus not part of this PR anymore

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -78,7 +78,7 @@
 #/avm/res/network/application-gateway/ @Azure/avm-res-network-applicationgateway-module-owners-bicep @Azure/avm-core-team-technical-bicep
 #/avm/res/network/application-gateway-web-application-firewall-policy/ @Azure/avm-res-network-applicationgatewaywebapplicationfirewallpolicy-module-owners-bicep @Azure/avm-core-team-technical-bicep
 /avm/res/network/application-security-group/ @Azure/avm-res-network-applicationsecuritygroup-module-owners-bicep @Azure/avm-core-team-technical-bicep
-#/avm/res/network/azure-firewall/ @Azure/avm-res-network-azurefirewall-module-owners-bicep @Azure/avm-core-team-technical-bicep
+/avm/res/network/azure-firewall/ @Azure/avm-res-network-azurefirewall-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
Comment (ChrisSidebotham) [inline .github/CODEOWNERS:81]: Blocking until Team Parent is configured

File: avm/res/network/azure-firewall/main.bicep [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+@description('Optional. Tier of an Azure Firewall.')
+@allowed([
+  'Basic'
+  'Standard'
+  'Premium'
+])
+param azureSkuTier string = 'Standard'
+
+@description('Conditional. Shared services Virtual Network resource ID. The virtual network ID containing AzureFirewallSubnet. If a Public IP is not provided, then the Public IP that is created as part of this module will be applied with the subnet provided in this variable. Required if `virtualHubId` is empty.')
+param vNetId string = ''
```
Comment (ChrisSidebotham) [inline avm/res/network/azure-firewall/main.bicep:17]:
```suggestion
param virtualNetworkId string = ''
```

**Review by ChrisSidebotham — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS (hunk 2 above)
Comment (rahalan) [reply .github/CODEOWNERS:81]: @matebarabas can you do that?

Comment (hundredacres) [reply .github/CODEOWNERS:81]: I've created both Teams:

https://github.com/orgs/Azure/teams/avm-res-network-azurefirewall-module-owners-bicep/
https://github.com/orgs/Azure/teams/avm-res-network-azurefirewall-module-contributors-bicep/

Comment (hundredacres) [reply .github/CODEOWNERS:81]: Parent teams?

Comment (rahalan) [reply .github/CODEOWNERS:81]: It is 'avm-technical-reviewers-bicep' for the owners team and 'avm-module-contributors-bicep' for the contributors team, see in the settings of the respective team: 
image

File: avm/res/network/azure-firewall/main.bicep [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+resource azureFirewall_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock ?? {}) && lock.?kind != 'None') {
+  name: lock.?name ?? 'lock-${name}'
+  properties: {
+    level: lock.?kind ?? ''
+    notes: lock.?kind == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot delete or modify the resource or child resources.'
+  }
+  scope: azureFirewall
+}
+
+resource azureFirewall_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for (diagnosticSetting, index) in (diagnosticSettings ?? []): {
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/main.bicep:246]: Uses outdated schema

File: avm/res/network/azure-firewall/main.bicep [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+  condition: string?
+
+  @description('Optional. Version of the condition.')
+  conditionVersion: '2.0'?
+
+  @description('Optional. The Resource Id of the delegated managed identity resource.')
+  delegatedManagedIdentityResourceId: string?
+}[]?
+
+type diagnosticSettingType = {
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/main.bicep:348]: Must be updated to latest schema

File: avm/res/network/azure-firewall/main.bicep [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    publicIpPrefixResourceId: contains(managementIPAddressObject, 'managementIPPrefixResourceId') ? (!(empty(managementIPAddressObject.publicIPPrefixResourceId)) ? managementIPAddressObject.publicIPPrefixResourceId : '') : ''
+    publicIPAllocationMethod: contains(managementIPAddressObject, 'managementIPAllocationMethod') ? (!(empty(managementIPAddressObject.publicIPAllocationMethod)) ? managementIPAddressObject.publicIPAllocationMethod : 'Static') : 'Static'
+    skuName: contains(managementIPAddressObject, 'skuName') ? (!(empty(managementIPAddressObject.skuName)) ? managementIPAddressObject.skuName : 'Standard') : 'Standard'
+    skuTier: contains(managementIPAddressObject, 'skuTier') ? (!(empty(managementIPAddressObject.skuTier)) ? managementIPAddressObject.skuTier : 'Regional') : 'Regional'
+    roleAssignments: contains(managementIPAddressObject, 'roleAssignments') ? (!empty(managementIPAddressObject.roleAssignments) ? managementIPAddressObject.roleAssignments : []) : []
+    diagnosticSettings: managementIPAddressObject.?diagnosticSettings
+    location: location
+    tags: managementIPAddressObject.?tags ?? tags
+    zones: zones
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/main.bicep:199]: Should allow configuration of like you did with tags

File: avm/res/network/azure-firewall/tests/e2e/addpip/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-network.azurefirewalls-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param resourceLocation string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'nafaddpip'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/tests/e2e/addpip/main.test.bicep:21]: should not be in this or any other test file

File: avm/res/network/azure-firewall/tests/e2e/addpip/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      publicIPAllocationMethod: 'Static'
+      roleAssignments: [
+        {
+          roleDefinitionIdOrName: 'Reader'
+          principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+          principalType: 'ServicePrincipal'
+        }
+      ]
+    }
+    tags: {
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/tests/e2e/addpip/main.test.bicep:78]: Not required for pip example

File: avm/res/network/azure-firewall/tests/e2e/custompip/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            }
+          ]
+          eventHubName: diagnosticDependencies.outputs.eventHubNamespaceEventHubName
+          eventHubAuthorizationRuleResourceId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+          storageAccountResourceId: diagnosticDependencies.outputs.storageAccountResourceId
+          workspaceResourceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+        }
+      ]
+    }
+    tags: {
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/tests/e2e/custompip/main.test.bicep:102]: not required for custom pip test

File: avm/res/network/azure-firewall/tests/e2e/hubcommon/main.test.bicep [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    enableTelemetry: enableTelemetry
+    name: '${namePrefix}${serviceShort}001'
+    firewallPolicyId: nestedDependencies.outputs.firewallPolicyResourceId
+    virtualHubId: nestedDependencies.outputs.virtualHubResourceId
+    hubIPAddresses: {
+      publicIPs: {
+        count: 1
+      }
+    }
+    tags: {
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/tests/e2e/hubcommon/main.test.bicep:67]: Not required for hub

File: avm/res/network/azure-firewall/tests/e2e/waf-aligned/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            category: 'AllMetrics'
+          }
+        ]
+        eventHubName: diagnosticDependencies.outputs.eventHubNamespaceEventHubName
+        eventHubAuthorizationRuleResourceId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+        storageAccountResourceId: diagnosticDependencies.outputs.storageAccountResourceId
+        workspaceResourceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+      }
+    ]
+    lock: {
```
Comment (AlexanderSehr) [inline avm/res/network/azure-firewall/tests/e2e/waf-aligned/main.test.bicep:142]: Not requried for WAF alignment

File: .github/CODEOWNERS (hunk 2 above)
Comment (hundredacres) [reply .github/CODEOWNERS:81]: Ok, updated.

File: avm/res/network/azure-firewall/main.bicep (hunk 4 above)
Comment (rahalan) [reply avm/res/network/azure-firewall/main.bicep:246]: done

File: avm/res/network/azure-firewall/main.bicep (hunk 6 above)
Comment (rahalan) [reply avm/res/network/azure-firewall/main.bicep]: done

File: avm/res/network/azure-firewall/tests/e2e/addpip/main.test.bicep (hunk 7 above)
Comment (rahalan) [reply avm/res/network/azure-firewall/tests/e2e/addpip/main.test.bicep]: done

File: avm/res/network/azure-firewall/tests/e2e/custompip/main.test.bicep (hunk 9 above)
Comment (rahalan) [reply avm/res/network/azure-firewall/tests/e2e/custompip/main.test.bicep]: done

**Review by AlexanderSehr — APPROVED** [review]

**Review by ChrisSidebotham — APPROVED** [review]

**Pull request merged**